### PR TITLE
[Relay][Pass] Don't consider constants as free vars in MergeComposite

### PR DESCRIPTION
**Pull request by soiferj**
Based on my post [here](https://discuss.tvm.ai/t/rfc-external-codegen-defining-composite-relay-operators/5470/9), if a pattern contains a constant, that constant should not be a parameter to the function that `MergeComposite` generates. This is because it is assumed that the constant is dealt with internally by the external codegen.

A summary of the fix is: create a new `const_list` which is passed around to `ExtractPattern` calls. When creating the final function, we call `GetFreeVarsWithoutConst` instead of `FreeVars`, as `FreeVars` also returns constants. `GetFreeVarsWithoutConst` will only return free vars that aren't also in `const_list`.

@comaniac @mbaret @zhiics would you be able to take a look?

### CODE REVIEW

File: src/relay/pass/merge_composite.cc [hunk 1]
```diff
@@ -182,6 +185,27 @@ class MergeCompositeWrapper : public ExprMutator {
...
 
  private:
+  /*! \brief Returns all free vars from expr that are not constants */
+  Array<Var> GetFreeVarsWithoutConst(const Expr& expr, const Array<Expr>& const_list) {
+    Array<Var> free_vars_without_const;
+    auto free_vars = FreeVars(expr);
+    for (const auto& free_var : free_vars) {
+      bool is_const = false;
+      for (const auto& const_var : const_list) {
+        if (free_var.get() == const_var.get()) {
```
Comment (zhiics) [inline src/relay/pass/merge_composite.cc]: just `free_var == const_var`?

File: src/relay/pass/merge_composite.cc [hunk 2]
```diff
@@ -162,9 +164,10 @@ class MergeCompositeWrapper : public ExprMutator {
     CHECK(pattern.defined());
     Map<std::string, Array<Expr>> args_map;
     Map<Expr, Expr> call_map;
-    auto extract = ExtractPattern(pattern, call, &args_map, &call_map);
+    Array<Expr> const_list;
+    auto extract = ExtractPattern(pattern, call, &args_map, &call_map, &const_list);
```
[the remark this answers is not in the document]
Comment (soiferj) [reply src/relay/pass/merge_composite.cc]: Updated

File: src/relay/pass/merge_composite.cc (hunk 1 above)
Comment (soiferj) [reply src/relay/pass/merge_composite.cc]: Updated

**Comment by mbaret** [comment]
I have a feeling you may be able to get around this problem by using bind_params_by_name. I have to do this when using merge composite because weight parameters are still treating as variables until this pass it called, at which time they are replaced with constants.

Could you try running:
```
f = relay.build_module.bind_params_by_name(mod["main"], params)
mod = tvm.IRModule()
mod["main"] = f
```
before the merge composite?

File: src/relay/pass/merge_composite.cc [hunk 3]
```diff
@@ -40,37 +40,34 @@ class MergeCompositeWrapper : public ExprMutator {
...
         return Expr();
       }
-      return (*var_map)[pattern->name_hint()][0];
+      return var_map_[pattern->name_hint()][0];
     }
   }
 
-  Expr ExtractPattern(const Constant& pattern, const Expr& root,
-          Map<std::string, Array<Expr>>* var_map) {
+  Expr ExtractPattern(const Constant& pattern, const Expr& root) {
```
Comment (zhiics) [inline src/relay/pass/merge_composite.cc]: I think this one can be safely removed

**Comment by soiferj** [comment]
@mbaret thanks for the heads-up. It looks like this works. Do you still think this PR is a worthwhile improvement?

**Comment by zhiics** [comment]
Thanks @mbaret. If that's the case, I think we can probably just add a unit test and apply the bind pass. All other changes could be reverted. How do you think?

**Comment by mbaret** [comment]
I think it actually implies a slightly different improvement should be made :) Currently no real checking happens in the ExtractPattern for constants, they are just always accepted as part of the pattern. Probably what should happen instead is that the pattern should fail to match if it's expecting a constant and instead sees a variable.

Adding a quick unit test to make sure it doesn't incorrectly match a constant to a var and then implementing that check in ExtractPattern should be sufficient.

**Comment by soiferj** [comment]
Sounds good. I'll close this PR for now.

**Comment by corehalt** [comment]
> I have a feeling you may be able to get around this problem by using bind_params_by_name. I have to do this when using merge composite because weight parameters are still treating as variables until this pass it called, at which time they are replaced with constants.
> 
> Could you try running:
> 
> ```
> f = relay.build_module.bind_params_by_name(mod["main"], params)
> mod = tvm.IRModule()
> mod["main"] = f
> ```
> 
> before the merge composite?

@mbaret @soiferj  I tried to call `bind_params_by_name()` before merge composite. 
Before, if I don't call `bind_params_by_name()` I get something like:
```
  %155 = nn.relu(%154) /* ty=Tensor[(1, 1024, 7, 7), float32] */;
  %157 = fn (%scompiler_input52: Tensor[(1, 1024, 7, 7), float32], %scompiler_input53: Tensor[(1024, 1024, 1, 1), float32], Compiler="scompiler", ExternalSymbol="scompiler_0", Primitive=1) -> Tensor[(1, 1024, 7, 7), float32] {
    %156 = fn (%x26: Tensor[(1, 1024, 7, 7), float32], %y26: Tensor[(1024, 1024, 1, 1), float32], Primitive=1, Composite="conv") -> Tensor[(1, 1024, 7, 7), float32] {
      nn.conv2d(%x26, %y26, padding=[0, 0, 0, 0], channels=1024, kernel_size=[1, 1]) /* ty=Tensor[(1, 1024, 7, 7), float32] */
    };
    %156(%scompiler_input52, %scompiler_input53) /* ty=Tensor[(1, 1024, 7, 7), float32] */
  };
  %158 = %157(%155, %separable_conv_block_13_conv2_weight) /* ty=Tensor[(1, 1024, 7, 7), float32] */;
```
But if I call `bind_params_by_name()` I can see something different:
```
  %129 = nn.relu(%128) /* ty=Tensor[(1, 1024, 7, 7), float32] */;
  %131 = fn (%scompiler_input52: Tensor[(1, 1024, 7, 7), float32], %scompiler_input53: Tensor[(1024, 1024, 1, 1), float32], Compiler="scompiler", ExternalSymbol="scompiler_0", Primitive=1) -> Tensor[(1, 1024, 7, 7), float32] {
    %130 = fn (%x26: Tensor[(1, 1024, 7, 7), float32], %y26: Tensor[(1024, 1024, 1, 1), float32], Primitive=1, Composite="conv") -> Tensor[(1, 1024, 7, 7), float32] {
      nn.conv2d(%x26, %y26, padding=[0, 0, 0, 0], channels=1024, kernel_size=[1, 1]) /* ty=Tensor[(1, 1024, 7, 7), float32] */
    };
    %130(%scompiler_input52, %scompiler_input53) /* ty=Tensor[(1, 1024, 7, 7), float32] */
  };
  %132 = %131(%129, meta[relay.Constant][52] /* ty=Tensor[(1024, 1024, 1, 1), float32] */ /* ty=Tensor[(1024, 1024, 1, 1), float32] */) /* ty=Tensor[(1, 1024, 7, 7), float32] */;
```
The thing is that later `meta[relay.Constant][52]` is exposed as a Relay.VarNode on the codegen anyway. Is there any way of accessing these constants from a custom codegen?

**Pull request closed without merging**